core/src/main/java/ai/timefold/solver/core/impl/move/streams/dataset/AbstractDatasetInstance.java

```diff
@@ -0,0 +1,37 @@
+package ai.timefold.solver.core.impl.move.streams.dataset;
+
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Random;
+
+import ai.timefold.solver.core.impl.bavet.common.tuple.AbstractTuple;
+import ai.timefold.solver.core.impl.bavet.common.tuple.TupleLifecycle;
+
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public abstract class AbstractDatasetInstance<Solution_, Tuple_ extends AbstractTuple>
+        implements TupleLifecycle<Tuple_> {
+
+    private final AbstractDataset<Solution_, Tuple_> parent;
+    protected final int inputStoreIndex;
+
+    public AbstractDatasetInstance(AbstractDataset<Solution_, Tuple_> parent, int inputStoreIndex) {
+        this.parent = Objects.requireNonNull(parent);
+        this.inputStoreIndex = inputStoreIndex;
+    }
+
+    public AbstractDataset<Solution_, Tuple_> getParent() {
+        return parent;
+    }
+
+    @Override
+    public void update(Tuple_ tuple) {
```

Comment:
Should we delegate the implementation to the child classes?

Reply:
The contract from `TupleLifecycle` includes the update method. Ignoring it does not seem correct to me because if it is called, then the caller expects the tuple to be updated.